dataset.py

```diff
@@ -108,17 +109,85 @@ def preprocess(self, do_val_split=True):
+            df_out = df_out.join(self._preprocess_deviceCategory(df))
+            df_out['geoNetwork.first_x_longitude'], df_out['geoNetwork.last_x_longitude'], df_out['geoNetwork.first_y_longitude'], df_out['geoNetwork.last_y_longitude'] = self._preprocess_longitudes_and_latitudes(df)
+            df_out = df_out.join(self._preprocess_country(df))
+            df_out = df_out.join(self._preprocess_metro(df))
             dfs_out.append((df_out, df_labels))
 
         return dfs_out
 
+    def _preprocess_country(self, df):
+        # One-hot encode the countries.
+        new_country_col = pd.Series(df['geoNetwork.country'])
+        country_dummies_df = pd.get_dummies(new_country_col)
+
+        return country_dummies_df
+
+    def _preprocess_metro(self, df):
+        # One-hot encode the metropolitan areas.
+        new_metro_col = pd.Series(df['geoNetwork.metro'])
+        metro_dummies_df = pd.get_dummies(new_metro_col)
+
+        return metro_dummies_df
+
+    def _preprocess_longitudes_and_latitudes(self, df):
+        """Preprocesses the columns u'geoNetwork.latitude',
+           u'geoNetwork.longitude', and u'geoNetwork.metro' in the dataset train.csv.
+           Creates columns of the first and last x and y longitudes and latitudes (4 columns) of each visitor.
+           Also creates dummy columns of the country and metropolitan area of each visit.
+           
+           Returns:
+               The training dataframe (indexed by visitor) with the added longitude/latitude columns.
+        """
+        
+        train_df = df.copy(deep=False)
+        
+        # Preprocess the numeric columns. Group by visitor and standardize, impute missing values, and normalize
+        
+        # Impute the missing values in latitudes and longitudes.
+        imp = impute.SimpleImputer(missing_values='not available in demo dataset', strategy='mean')
+        train_df['imputed_latitude'] = imp.transform(train_df['geoNetwork.latitude'])
+        train_df['imputed_longitude'] = imp.transform(train_df['geoNetwork.longitude'])
+        
+        # Convert longitude and latitude into x and y coordinates.
+        # First convert from degrees to radians, then take sin and cosine.
+        train_df['x_longitude'] = train_df['geoNetwork.longitude'] * (math.pi / 180)
+        train_df['x_longitude'] = numpy.cos(train_df['x_longitude'])
+        train_df['y_longitude'] = train_df['geoNetwork.longitude'] * (math.pi / 180)
+        train_df['y_longitude'] = numpy.sin(train_df['y_longitude'])
+        
+        train_df['x_latitude'] = train_df['geoNetwork.latitude'] * (math.pi / 180)
+        train_df['x_latitude'] = numpy.cos(train_df['x_latitude'])
+        train_df['y_latitude'] = train_df['geoNetwork.latitude'] * (math.pi / 180)
+        train_df['y_latitude'] = numpy.sin(train_df['y_latitude'])
+        
+        # Sort by date.
+        train_df = train_df.sort_values(by = ['date'])
```

Comment:
check that "date" is sorted properly. I don't believe that this column is converted to datetime format at this point, so the date is probably just a string which might not be sorted correctly

Reply:
date is of type int64.

Reply:
And the default for sort_by() is ascending, so the oldest dates should appear first and the newest dates last.

Reply:
Regardless, you could have a problem here when you sort. For example, if the dates are represented as `MMDDYYYY`, which is pretty common. In this case:

November 26 2018 = 11262018
December 26 2017 = 12262017

If you sort in ascending order, the 2018 date will still come first, even though it happened last. The best way to ensure that the dates are sorted properly is to check their representation, convert them to a datetime object, sort, and then convert back to an integer. The integer representation should be something with a sensible ordering, like "# of seconds since january 1 2017" (https://en.wikipedia.org/wiki/Unix_time)

Reply:
I understand

Reply:
But the dates are YYYYMMDD, so that shouldn't be an issue

Reply:
![screen shot 2018-11-26 at 4 18 48 pm](https://user-images.githubusercontent.com/23203199/49042751-03265680-f197-11e8-91f0-287a401fb619.png)

Reply:
Ok, sounds good.